### PR TITLE
[!feat] Refactor Okx

### PR DESCRIPTION
**Pull request by colmazia**
## Please ensure the following requirements are met before submitting a pull request:

- [x] The pull request is targeted against the correct target branch
- [ ] The pull request is linked to an issue with appropriate discussion and an accepted design OR is linked to a spec that describes the work.
- [ ] The pull request includes a description of the implementation/work done in detail.
- [x] The pull request includes any appropriate unit/integration tests
- [ ] You have added a relevant changelog entry to `CHANGELOG.md`
- [x] You have re-reviewed the files affected by the pull request (e.g. using the `Files changed` tab in the GitHub PR explorer)

### CODE REVIEW

File: bothan-okx/src/api/types/channel.rs [hunk 1]
```diff
@@ -1,7 +1,7 @@
 use serde::{Deserialize, Serialize};
 
 /// Represents a response from a channel subscription.
-#[derive(Clone, Debug, Serialize, Deserialize)]
+#[derive(Clone, Debug, Serialize, Deserialize, PartialEq)]
```
Comment (warittornc) [inline bothan-okx/src/api/types/channel.rs:4]: ditto

File: Cargo.toml [hunk 2]
```diff
@@ -7,7 +7,15 @@ exclude = [".github/"]
...
-members = ["bothan-core", "bothan-binance", "bothan-coingecko", "bothan-kraken", "bothan-api/server", "bothan-api/server-cli"]
+members = [
+    "bothan-core",
+    "bothan-binance",
+    "bothan-coingecko",
+    "bothan-kraken",
+    "bothan-okx",
+    "bothan-api/server",
+    "bothan-api/server-cli",
+]
```
[the remark this answers is not in the document]
Comment (colmazia) [reply Cargo.toml]: fixed

File: bothan-kraken/src/worker/asset_worker.rs [hunk 3]
```diff
@@ -133,7 +133,11 @@ fn parse_ticker(ticker: TickerResponse) -> Result<AssetInfo, WorkerError> {
     let id = ticker.symbol.clone();
     let price_value =
         Decimal::from_f64(ticker.last).ok_or(WorkerError::InvalidPrice(ticker.last))?;
-    Ok(AssetInfo::new(id, price_value, 0))
+    Ok(AssetInfo::new(
+        id,
+        price_value,
+        chrono::Utc::now().timestamp(),
+    ))
```
[the remark this answers is not in the document]
Comment (colmazia) [reply bothan-kraken/src/worker/asset_worker.rs:140]: fixed

File: bothan-okx/src/api/types.rs [hunk 4]
```diff
@@ -1,15 +1,15 @@
...
+pub use message::WebSocketMessageResponse;
 
 pub mod channel;
 pub mod message;
 
 pub const DEFAULT_URL: &str = "wss://ws.okx.com:8443/ws/v5/public";
 
 /// Represents the different types of responses from the OKX WebSocket API.
-#[derive(Clone, Debug, Serialize, Deserialize)]
+#[derive(Clone, Debug, Serialize, Deserialize, PartialEq)]
```
[the remark this answers is not in the document]
Comment (colmazia) [reply bothan-okx/src/api/types.rs]: fixed

File: bothan-okx/src/api/types/message.rs [hunk 5]
```diff
@@ -26,7 +26,7 @@ pub struct WebSocketMessage<T> {
 }
 
 /// Represents a response to a WebSocket message.
-#[derive(Clone, Debug, Serialize, Deserialize)]
+#[derive(Clone, Debug, Serialize, Deserialize, PartialEq)]
```
[the remark this answers is not in the document]
Comment (colmazia) [reply bothan-okx/src/api/types/message.rs]: fixed

File: bothan-okx/src/api/websocket.rs [hunk 6]
```diff
@@ -50,45 +51,58 @@ impl OkxWebSocketConnection {
...
+    pub async fn subscribe_ticker(&mut self, inst_ids: &[&str]) -> Result<(), SendError> {
         let ticker_args = build_ticker_arguments(inst_ids);
         let msg = WebSocketMessage {
             op: Op::Subscribe,
             args: Some(ticker_args),
         };
-        let message = Message::Text(serde_json::to_string(&msg)?);
+
+        // Create the subscription payload.
+        let payload = json!(msg);
```
[the remark this answers is not in the document]
Comment (colmazia) [reply bothan-okx/src/api/websocket.rs]: fixed

File: bothan-okx/src/api/websocket.rs [hunk 7]
```diff
@@ -50,45 +51,58 @@ impl OkxWebSocketConnection {
...
-    pub async fn unsubscribe_ticker(&mut self, inst_ids: &[&str]) -> Result<(), Error> {
+    pub async fn unsubscribe_ticker(&mut self, inst_ids: &[&str]) -> Result<(), SendError> {
         let ticker_args = build_ticker_arguments(inst_ids);
         let msg = WebSocketMessage {
             op: Op::Unsubscribe,
             args: Some(ticker_args),
         };
-        let message = Message::Text(serde_json::to_string(&msg)?);
+        // Create the unsubscription payload.
+        let payload = json!(msg);
```
[the remark this answers is not in the document]
Comment (colmazia) [reply bothan-okx/src/api/websocket.rs]: fixed

File: bothan-okx/src/api/types/channel.rs (hunk 1 above)
Comment (colmazia) [reply bothan-okx/src/api/types/channel.rs]: fixed

File: Cargo.toml [hunk 8]
```diff
@@ -7,7 +7,15 @@ exclude = [".github/"]
 
 
 [workspace]
-members = ["bothan-core", "bothan-binance", "bothan-coingecko", "bothan-kraken", "bothan-api/server", "bothan-api/server-cli"]
+members = [
+    "bothan-api/server",
+    "bothan-api/server-cli",
+    "bothan-binance",
+    "bothan-core",
+    "bothan-coingecko",
```
Comment (warittornc) [inline Cargo.toml:15]: bothan-core should be after bothan-coingecko?

**Review by warittornc — APPROVED** [review]
after resolving the comment below, lgtm

File: Cargo.toml (hunk 8 above)
Comment (colmazia) [reply Cargo.toml:15]: so how should we order them, at first I just use your ordering then it is not your liking so I just went with alphabetically.
So please provide me the format you would like.

Comment (colmazia) [reply Cargo.toml:15]: we can fix this in the later source krub

Comment (warittornc) [reply Cargo.toml:15]: it should be alphabetically, "i" comes before "r"

Comment (colmazia) [reply Cargo.toml:15]: oh I remember I put it after, maybe I just blur

Comment (colmazia) [reply Cargo.toml]: fixed

File: bothan-okx/src/worker/error.rs [hunk 9]
```diff
@@ -0,0 +1,16 @@
+use crate::api;
+use bothan_core::store;
+use thiserror::Error;
+
+#[derive(Error, Debug)]
+pub(crate) enum WorkerError {
+    #[error("value is not a valid decimal: {0}")]
+    Underflow(#[from] rust_decimal::Error),
```
Comment (RogerKSI) [inline bothan-okx/src/worker/error.rs:8]: Underflow is just one of the decimal error? should it be InvalidDecimal instead?

Comment (colmazia) [reply bothan-okx/src/worker/error.rs]: fixed

**Review by RogerKSI — APPROVED** [review]

**Pull request merged**